File: kh1_src/tabs/journal_tabs/characters.py
from dash import Dash, html, dcc, callback, Input, Output, State, ALL
import kh1_src.kh1_utils as utils

@callback(
    Output("JournalCharactersDiv", "children"),
    Input("JournalCharactersTabs", "value"),
)
def __create_journal_characters_tabs(tab):
    if tab == "Characters 1":
        return __create_characters_1()
    if tab == "Characters 2":
        return __create_characters_2()
    if tab == "The Heartless":
        return __create_the_heartless()

def __create_characters_1():
    kh1 = utils.kh1
    chars1 = html.Div([
        html.Div([
            dcc.Checklist(
                options=[
                    {"label": k, "value": (1 << v % 16)}
                ],
                value=[kh1.journal_chars[v // 16] & (1 << v % 16)],
                id={"type": "Journal Character Entry", "index": v},
                style={"margin-top": 15},
            )
        ]) if type(v) == int else\
        html.Div([
            dcc.Markdown(k+":"),
            dcc.RadioItems(
                options=[
                    # This works because multipart entries always on different bits.
                    {"label": l, "value": (1 << p % 16)}\
                    for l, p in v.items()
                ],
                value=sum([kh1.journal_chars[p // 16] & (1 << p % 16) for p in v.values()]),
                id={"type": "Journal Character Multi Entry", "index": k},
            ),
        ])\
        for k, v in kh1.journal_chars_1_dict.items()
    ],
        style={"margin-top": 20},
    )
    return chars1

def __create_characters_2():
    kh1 = utils.kh1
    chars2 = html.Div([
        html.Div([
            dcc.Checklist(
                options=[
                    {"label": k, "value": (1 << v % 16)}
                ],
                value=[kh1.journal_chars[v // 16] & (1 << v % 16)],
                id={"type": "Journal Character Entry", "index": v},
                style={"margin-top": 15},
            )
        ]) if type(v) == int else\
        html.Div([
            dcc.Markdown(k+":"),
            dcc.RadioItems(
                options=[
                    # This works because multipart entries always on different bits.
                    {"label": l, "value": (1 << p % 16)}\
                    for l, p in v.items()
                ],
                value=sum([kh1.journal_chars[p // 16] & (1 << p % 16) for p in v.values()]),
                id={"type": "Journal Character Multi Entry", "index": k},
            ),
        ])\
        for k, v in kh1.journal_chars_2_dict.items()
    ],
        style={"margin-top": 20},
    )
    return chars2

def __create_the_heartless():
    kh1 = utils.kh1
    heartless_dict = kh1.heartless_fm_dict if kh1.fm else kh1.heartless_dict
    heartless = html.Div([
        html.Div([
            html.Label(k + ": "),
            dcc.Input(
                id={"type": "Heartless Kill Count", "index": v},
                type="number",
                value=kh1.heartless[v],
                min=0,
                max=0xFFFF,
                step=1,
                style={"width": 50},
            ),
        ],
            style={"margin-top": 20, "gap": 10},
        ) for k, v in heartless_dict.items()
    ])
    bosses = html.Div([
        dcc.Checklist(
            options=[
                {"label": k, "value": (1 << v % 16)}
            ],
            value=[kh1.journal_chars[v // 16] & (1 << v % 16)],
            id={"type": "Journal Character Entry", "index": v},
        ) for k, v in kh1.journal_boss_dict.items()
    ])
    return html.Div([
        html.H3("Heartless Kill Counts:"),
        heartless,
        html.H3("Heartless Bosses:"),
        bosses,
    ])

def create_journal_characters():
    jctabs = dcc.Tabs(id="JournalCharactersTabs", value="Characters 1")
    jctabs.children = [
        dcc.Tab(label="Characters 1", value="Characters 1"),
        dcc.Tab(label="Characters 2", value="Characters 2"),
        dcc.Tab(label="The Heartless", value="The Heartless"),
    ]
    return html.Div([
        jctabs,
        html.Div(id="JournalCharactersDiv"),
    ])

@callback(
    Input({"type": "Heartless Kill Count", "index": ALL}, "value"),
    State({"type": "Heartless Kill Count", "index": ALL}, "id"),
)
def the_heartless_callback(values, ids):
    kh1 = utils.kh1
    try:
        i = 0
        for id in ids:
            idx = id["index"]
            kh1.heartless[idx] = values[i]
            i += 1
    except:
        pass

@callback(
    Input({"type": "Journal Character Entry", "index": ALL}, "value"),
    State({"type": "Journal Character Entry", "index": ALL}, "id"),
)
def journal_characters_callback(values, ids):
    kh1 = utils.kh1
    for i in range(len(values)):
        v = ids[i]["index"]
        if (1 << v % 16) in values[i]:
            kh1.journal_chars[v // 16] |= (1 << v % 16)
        else:
            kh1.journal_chars[v // 16] &= ~(1 << v % 16)

@callback(
    Input({"type": "Journal Character Multi Entry", "index": ALL}, "value"),
    State({"type": "Journal Character Multi Entry", "index": ALL}, "id"),
)
def journal_characters_callback(values, ids):
    kh1 = utils.kh1
    for value, id in zip(values, ids):
        name = id["index"]
        if name in kh1.journal_chars_1_dict:
            for v in kh1.journal_chars_1_dict[name].values():
                if (1 << v % 16) == value:
                    kh1.journal_chars[v // 16] |= (1 << v % 16)
                else:
                    kh1.journal_chars[v // 16] &= ~(1 << v % 16)
        if name in kh1.journal_chars_2_dict:
            for v in kh1.journal_chars_2_dict[name].values():
                if (1 << v % 16) == value:
                    kh1.journal_chars[v // 16] |= (1 << v % 16)
                else:
                    kh1.journal_chars[v // 16] &= ~(1 << v % 16)

File: kh1_src/tabs/journal_tabs/characters.js
export function create_journal_characters() {
    const kh1jdiv = document.getElementById("kh1jdiv");
    kh1jdiv.innerHTML = `
    <div>
        <div>
            <h3>Entry:</h3>
            <select id="kh1jchartabs">
                <option value="Characters 1">Characters 1</option>
                <option value="Characters 2">Characters 2</option>
                <option value="The Heartless">The Heartless</option>
            </select>
        </div>
        <div id="kh1jchardiv"></div>
    </div>`;
    const kh1jchartabs = document.getElementById("kh1jchartabs");
    function tab_sel() {
        switch (kh1jchartabs.value) {
            case "Characters 1": {
                create_journal_chars_1();
                break;
            }
            case "Characters 2": {
                create_journal_chars_2();
                break;
            }
            case "The Heartless": {
                create_heartless();
                break;
            }
        }
    }
    kh1jchartabs.addEventListener("change", tab_sel);
    tab_sel();
}

function create_journal_chars_1() {
    const kh1jchardiv = document.getElementById("kh1jchardiv");
    let journal_chars_1 = `<div id="journal_chars_1">`;
    Object.entries(window.kh1.journal_chars_1_dict).forEach(([label, value]) => {
        if (typeof value == "number") {
            const idx = Math.floor(value / 16);
            const bit = value % 16;
            journal_chars_1 += `
            <label style="display: flex; align-items: center">
                <input
                    type="checkbox"
                    value=${value}
                    ${window.kh1.journal_chars[idx] & (1 << bit) ? "checked" : ""}
                >
                ${label}
            </label>`;
        }
        else {
            journal_chars_1 += `
            <h3>${label}</h3>
            <label style="display: flex; align-items: center">
                <input
                    type="radio"
                    name=${label}
                    value=0
                    checked
                >
                Locked
            </label>`;
            Object.entries(value).forEach(([k, v]) => {
                const idx = Math.floor(v / 16);
                const bit = v % 16;
                journal_chars_1 += `
                <label style="display: flex; align-items: center">
                    <input
                        type="radio"
                        name=${label}
                        value=${v}
                        ${window.kh1.journal_chars[idx] & (1 << bit) ? "checked" : ""}
                    >
                    ${k}
                </label>`;
            });
        }
    });
    journal_chars_1 += "</div>";
    kh1jchardiv.innerHTML = `
    <div>
        ${journal_chars_1}
    </div>`;
    journal_chars_1_callbacks();
}

function journal_chars_1_callbacks() {
    const journal_chars_1 = document.getElementById("journal_chars_1");
    journal_chars_1.addEventListener("change", (e) => {
        if (e.target.type == "checkbox") {
            const idx = Math.floor(e.target.value / 16);
            const bit = e.target.value % 16;
            if (e.target.checked)
                window.kh1.journal_chars[idx] |= (1 << bit);
            else
                window.kh1.journal_chars[idx] &= ~(1 << bit);
        }
        else {
            Object.entries(window.kh1.journal_chars_1_dict[e.target.name]).forEach(([label, value]) => {
                const idx = Math.floor(value / 16);
                const bit = value % 16;
                if (e.target.value == value)
                    window.kh1.journal_chars[idx] |= (1 << bit);
                else
                    window.kh1.journal_chars[idx] &= ~(1 << bit);
            });
        }
    });
}

function create_journal_chars_2() {
    const kh1jchardiv = document.getElementById("kh1jchardiv");
    let journal_chars_2 = `<div id="journal_chars_2">`;
    Object.entries(window.kh1.journal_chars_2_dict).forEach(([label, value]) => {
        if (typeof value == "number") {
            const idx = Math.floor(value / 16);
            const bit = value % 16;
            journal_chars_2 += `
            <label style="display: flex; align-items: center">
                <input
                    type="checkbox"
                    value=${value}
                    ${window.kh1.journal_chars[idx] & (1 << bit) ? "checked" : ""}
                >
                ${label}
            </label>`;
        }
        else {
            journal_chars_2 += `
            <h3>${label}</h3>
            <label style="display: flex; align-items: center">
                <input
                    type="radio"
                    name=${label}
                    value=0
                    checked
                >
                Locked
            </label>`;
            Object.entries(value).forEach(([k, v]) => {
                const idx = Math.floor(v / 16);
                const bit = v % 16;
                journal_chars_2 += `
                <label style="display: flex; align-items: center">
                    <input
                        type="radio"
                        name=${label}
                        value=${v}
                        ${window.kh1.journal_chars[idx] & (1 << bit) ? "checked" : ""}
                    >
                    ${k}
                </label>`;
            });
        }
    });
    journal_chars_2 += "</div>";
    kh1jchardiv.innerHTML = `
    <div>
        ${journal_chars_2}
    </div>`;
    journal_chars_2_callbacks();
}

function journal_chars_2_callbacks() {
    const journal_chars_2 = document.getElementById("journal_chars_2");
    journal_chars_2.addEventListener("change", (e) => {
        if (e.target.type == "checkbox") {
            const idx = Math.floor(e.target.value / 16);
            const bit = e.target.value % 16;
            if (e.target.checked)
                window.kh1.journal_chars[idx] |= (1 << bit);
            else
                window.kh1.journal_chars[idx] &= ~(1 << bit);
        }
        else {
            Object.entries(window.kh1.journal_chars_2_dict[e.target.name]).forEach(([label, value]) => {
                const idx = Math.floor(value / 16);
                const bit = value % 16;
                if (e.target.value == value)
                    window.kh1.journal_chars[idx] |= (1 << bit);
                else
                    window.kh1.journal_chars[idx] &= ~(1 << bit);
            });
        }
    });
}

function create_heartless() {
    const kh1jchardiv = document.getElementById("kh1jchardiv");
    const heartless_dict = window.kh1.fm ? window.kh1.heartless_fm_dict : window.kh1.heartless_dict;
    const heartless = `
    <div id="heartless">
        ${
            Object.entries(heartless_dict)
                .map(([label, value]) => `
                <label style="display: flex; gap: 10px; align-items: center">
                    ${label}:
                    <input
                        type="number"
                        name=${value}
                        min=0
                        max=9999
                        step=1
                        value=${window.kh1.heartless[value]}
                    >
                </label>`)
                .join("")
        }
    </div>`;
    const bosses = `
    <div id="bosses">
        ${
            Object.entries(window.kh1.journal_boss_dict)
                .map(([label, value]) => `
                <label style="display: flex; align-items: center">
                    <input
                        type="checkbox"
                        value=${value}
                        ${window.kh1.journal_chars[Math.floor(value / 16)] & (1 << value % 16) ? "checked" : ""}
                    >
                    ${label}
                </label>`)
                .join("")
        }
    </div>`;
    kh1jchardiv.innerHTML = `
    <div>
        <h3>Heartless Kill Counts:</h3>
        ${heartless}
        <h3>Heartless Bosses:</h3>
        ${bosses}
    </div>`;
    heartless_callbacks();
}

function heartless_callbacks() {
    const heartless = document.getElementById("heartless");
    heartless.addEventListener("change", (e) => {
        if (e.target.validity.valid)
            window.kh1.heartless[e.target.name] = e.target.value;
        e.target.value = window.kh1.heartless[e.target.name];
    });
    const bosses = document.getElementById("bosses");
    bosses.addEventListener("change", (e) => {
        const idx = Math.floor(e.target.value / 16);
        const bit = e.target.value % 16;
        if (e.target.checked)
            window.kh1.journal_chars[idx] |= (1 << bit);
        else
            window.kh1.journal_chars[idx] &= ~(1 << bit);
    });
}
